src/analyzer.ts

```diff
@@ -116,18 +123,66 @@ export class Analyzer {
     // re-analyze a particular file. Editors can use a UrlLoader that reads from
     // it's internal buffers.
     if (contents != null) {
+      this._scannedDocumentPromises.delete(resolvedUrl);
       this._scannedDocuments.delete(resolvedUrl);
-      this._parsedDocuments.delete(resolvedUrl);
+      this._parsedDocumentPromises.delete(resolvedUrl);
+      this._analyzedDocuments.delete(resolvedUrl);
+      this._analyzedDocumentPromises.delete(resolvedUrl);
+    }
+
+    const cachedResult = this._analyzedDocumentPromises.get(resolvedUrl);
+    if (cachedResult) {
+      return cachedResult;
     }
+    // TODO(justinfagnani): also check _analyzedDocuments?
+
+    const promise = (async() => {
+      // Make sure we wait and return a Promise before doing any work, so that
+      // the Promise is cached before anything else happens.
+      await Promise.resolve();
+      const scannedDocument = await this._scan(resolvedUrl, contents);
+      return this._makeDocument(scannedDocument);
+    })();
+    this._analyzedDocumentPromises.set(resolvedUrl, promise);
```

Comment:
Yeah ideally imports would be able to influence the scanned document that's produced..

Hm.